x-pack/dockerlogbeat/magefile.go

```diff
@@ -47,132 +57,289 @@ func getPluginName() (string, error) {
+// createContainer builds the plugin and creates the container that will later become the rootfs used by the plugin
+func createContainer(cli *client.Client) error {
+	goVersion, err := mage.GoVersion()
+	if err != nil {
+		return errors.Wrap(err, "error determining go version")
+	}
 
 	dockerLogBeatDir, err := os.Getwd()
 	if err != nil {
 		return errors.Wrap(err, "error getting work dir")
 	}
 
+	if !strings.Contains(dockerLogBeatDir, "dockerlogbeat") {
+		return errors.Errorf("not in dockerlogbeat directory: %s", dockerLogBeatDir)
+	}
+
+	// change back to the root beats dir so we can send the proper build context to docker
 	err = os.Chdir("../..")
 	if err != nil {
 		return errors.Wrap(err, "error changing directory")
 	}
 
-	gv, err := mage.GoVersion()
+	// start to build the root container that'll be used to build the plugin
+	tmpDir, err := ioutil.TempDir("", "dockerBuildTar")
 	if err != nil {
-		return errors.Wrap(err, "error determining go version")
+		return errors.Wrap(err, "Error locating temp dir")
+	}
+	tarPath := filepath.Join(tmpDir, "tarRoot.tar")
+	err = sh.RunV("tar", "cf", tarPath, "./")
+	if err != nil {
+		return errors.Wrap(err, "error creating tar")
 	}
 
-	err = sh.RunV("docker", "build", "--build-arg", "versionString"+"="+gv, "--target", "final", "-t", "rootfsimage", "-f", "x-pack/dockerlogbeat/Dockerfile", ".")
+	buildContext, err := os.Open(tarPath)
 	if err != nil {
+		return errors.Wrap(err, "error opening temp dur")
+	}
+	defer buildContext.Close()
+
+	buildOpts := types.ImageBuildOptions{
+		BuildArgs:  map[string]*string{"versionString": &goVersion},
+		Target:     "final",
+		Tags:       []string{rootImageName},
+		Dockerfile: "x-pack/dockerlogbeat/Dockerfile",
+	}
+	//build, wait for output
+	buildResp, err := cli.ImageBuild(context.Background(), buildContext, buildOpts)
+	defer buildResp.Body.Close()
+	// buf := new(bytes.Buffer)
+	// _, errBufRead := buf.ReadFrom(buildResp.Body)
+	buf, errBufRead := ioutil.ReadAll(buildResp.Body)
+	if errBufRead != nil {
+		return errors.Wrap(err, "error reading from docker output")
+	}
+	if err != nil {
+		fmt.Printf("Docker response: \n %s\n", string(buf))
 		return errors.Wrap(err, "error building final container image")
 	}
 
+	// move back to the x-pack dir
 	err = os.Chdir(dockerLogBeatDir)
 	if err != nil {
 		return errors.Wrap(err, "error returning to dockerlogbeat dir")
 	}
 
-	os.Mkdir(filepath.Join(packageStagingDir, "rootfs"), 0755)
+	err = sh.Rm(tmpDir)
+	if err != nil {
+		return errors.Wrap(err, "error removing temp dir")
+	}
+
+	return nil
+}
+
+// Build builds docker rootfs container root
+// There's a somewhat complicated process for this:
+// * Create a container to build the plugin itself
+// * Copy that to a bare-bones container that will become the runc container used by docker
+// * Export that container
+// * Unpack the tar from the exported container
+// * send this to the plugin create API endpoint
+func Build() error {
```

Comment:
I think you can define mage targets with a context and then you can reuse this context in all the requests to the docker API.
```suggestion
func Build(ctx context.Context) error {
```